### ISSUE TITLE
Concurrent calls to collect stats on Windows using containerd hangs the metrics endpoint

### ISSUE THREAD

**Issue by jsturtevant** [kubernetes/kubernetes#104111]
<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


#### What happened:

Concurrent calls to collect stats  on Windows using containerd hangs the metrics endpoint.  

#### What you expected to happen:

Should be able to call metric endpoint (used by metric server)

#### How to reproduce it (as minimally and precisely as possible):

open a proxy to the node: `kubectl proxy`

run the program from https://gist.github.com/jsturtevant/c6c35378617073c0b9a647207a52b8ec which calls the metric endpoint 10 times consecutively:

```
go run . -l 10 -n 3809k8s01000000 -s 0 
Starting 0 containers on node 3809k8s01000000
pods at start: 1
starting 0 containers
run call to stats 10 times
timing: http://localhost:8001/api/v1/nodes/3809k8s01000000/proxy/stats/summary?only_cpu_and_memory=true took 495.574231ms
timing: http://localhost:8001/api/v1/nodes/3809k8s01000000/proxy/stats/summary?only_cpu_and_memory=true took 2m51.297855508s
timing: http://localhost:8001/api/v1/nodes/3809k8s01000000/proxy/stats/summary?only_cpu_and_memory=true took 2m51.297848339s
timing: http://localhost:8001/api/v1/nodes/3809k8s01000000/proxy/stats/summary?only_cpu_and_memory=true took 2m51.297942525s
timing: http://localhost:8001/api/v1/nodes/3809k8s01000000/proxy/stats/summary?only_cpu_and_memory=true took 2m51.298048848s

<hangs>
```

look at the metric server logs on the cluster and see context deadline exceeded:

```
k logs metrics-server-67b6bdd56c-gxrc4

E0803 22:37:08.994511       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.240.0.4:10250/stats/summary?only_cpu_and_memory=true\": context deadline exceeded" node="3809k8s01000000"                                                                                                                                                                               
E0803 22:37:23.994176       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.240.0.4:10250/stats/summary?only_cpu_and_memory=true\": context deadline exceeded" node="3809k8s01000000"                                                                                                                                                                               
E0803 22:37:38.993239       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.240.0.4:10250/stats/summary?only_cpu_and_memory=true\": context deadline exceeded" node="3809k8s01000000"                                                                                                                                                                               
E0803 22:37:53.993852       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.240.0.4:10250/stats/summary?only_cpu_and_memory=true\": context deadline exceeded" node="3809k8s01000000"                                                                                                                                                                               
E0803 22:38:08.994044       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.240.0.4:10250/stats/summary?only_cpu_and_memory=true\": context deadline exceeded" node="3809k8s01000000" 
```

#### Anything else we need to know?:

Restarting containerd and kublet on the node will fix the issue.

Dumping the kubelet go routines will show it waiting for calls to 

```
curl http://localhost:8001/api/v1/nodes/3809k8s01000000/proxy/debug/pprof/goroutine\?debug\=2 | ag waitonheader -A 30
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current              
                                 Dload  Upload   Total   Spent    Left  Speed                
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0        
             
k8s.io/kubernetes/vendor/google.golang.org/grpc/internal/transport.(*Stream).waitOnHeader(0xc002277a00)                                                                                                                                                                                                                                                                             
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/internal/transport/transport.go:318 +0xa5                                                                                                                                                                                                                               
k8s.io/kubernetes/vendor/google.golang.org/grpc/internal/transport.(*Stream).RecvCompress(...)                                                                                                                                                                                                                                                                                      
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/internal/transport/transport.go:333                                                                                                                                                                                                                                     
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*csAttempt).recvMsg(0xc0025f2580, 0x56ea8a0, 0xc00088eba0, 0x0, 0x0, 0x0)                                                                                                                                                                                                                                                          
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/stream.go:871 +0x765                                                                                                                                                                                                                                                    
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*clientStream).RecvMsg.func1(0xc0025f2580, 0xc00221c478, 0x2)                                                                                                                                                                                                                                                                      
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/stream.go:736 +0x4d                                                                                                                                                                                                                                                     
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*clientStream).withRetry(0xc0027679e0, 0xc0019c03c8, 0xc0019c0398, 0xc00221c47a, 0xc002d09b90)                                                                                                                                                                                                                                     
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/stream.go:594 +0xa6                                                                                                                                                                                                                                                     
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*clientStream).RecvMsg(0xc0027679e0, 0x56ea8a0, 0xc00088eba0, 0x0, 0x0)                                                                                                                                                                                                                                                            
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/stream.go:735 +0x10d                                                                                                                                                                                                                                                    
k8s.io/kubernetes/vendor/google.golang.org/grpc.invoke(0x5fce8f0, 0xc0008f9f80, 0x59be4c7, 0x33, 0x56ea760, 0xc000f18f70, 0x56ea8a0, 0xc00088eba0, 0xc000236a80, 0xc0002839c0, ...)                                                                                                                                                                                                 
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/call.go:73 +0x14f                                                                                                                                                                                                                                                       
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*ClientConn).Invoke(0xc000236a80, 0x5fce8f0, 0xc0008f9f80, 0x59be4c7, 0x33, 0x56ea760, 0xc000f18f70, 0x56ea8a0, 0xc00088eba0, 0x0, ...)                                                                                                                                                                                            
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/call.go:37 +0x1c5                                                                                                                                                                                                                                                       
k8s.io/kubernetes/vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2.(*runtimeServiceClient).ListContainerStats(0xc000007768                                                                                                                                                      
```

Dumping containerd routines shows it hangs as well:

```
ctr pprof goroutines | sls -Pattern dispatch -Context 0, 30

github.com/containerd/ttrpc.(*Client).dispatch(0xc000440580, 0x22bf240, 0xc0008909c0, 0xc0005dc360, 0xc0000dd000, 0xc0009e9120, 0x5)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/containerd/ttrpc/client.go:155 +0x18c
github.com/containerd/ttrpc.defaultClientInterceptor(0x22bf240, 0xc0008909c0, 0xc0005dc360, 0xc0000dd000, 0xc0009e9110, 0xc0009e9120, 0x1e, 0x2030002)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/containerd/ttrpc/interceptor.go:49 +0x55
github.com/containerd/ttrpc.(*Client).Call(0xc000440580, 0x22bf240, 0xc0008909c0, 0x20bc79a, 0x17, 0x20a1a36, 0x5, 0x2009c80, 0xc000890b70, 0x2009dc0, ...)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/containerd/ttrpc/client.go:132 +0x33a
github.com/containerd/containerd/runtime/v2/task.(*taskClient).Stats(0xc000006e78, 0x22bf240, 0xc0008909c0, 0xc000890b70, 0xf, 0xd6c3ad, 0x6109a3fb)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/runtime/v2/task/shim.pb.go:3697 +0xc5
github.com/containerd/containerd/runtime/v2.(*shim).Stats(0xc0002449c0, 0x22bf240, 0xc0008909c0, 0x1, 0x1, 0xc00008cd30)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/runtime/v2/shim.go:460 +0xa7
github.com/containerd/containerd/services/tasks.getTasksMetrics(0x22bf240, 0xc0008909c0, 0x2290740, 0xc0005d8330, 0xc000e21400, 0x1e, 0x20, 0xc0006596c0)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/services/tasks/local.go:623 +0x1aa
github.com/containerd/containerd/services/tasks.(*local).Metrics(0xc0003907d0, 0x22bf240, 0xc0008909c0, 0xc0006590c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000c45400)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/services/tasks/local.go:580 +0x17b
github.com/containerd/containerd/pkg/cri/server.(*criService).ListContainerStats(0xc00043d500, 0x22bf240, 0xc0008909c0, 0xc0009e8db0, 0x22bf240, 0xc0008909c0, 0x1)
        D:/a/containerd/containerd/src/github.com/containerd/containerd/pkg/cri/server/container_stats_list.go:38 +0x1b3
```

#### Environment:
- Kubernetes version (use `kubectl version`): 1.21.2
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Windows
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:
  - containerd: 1.5.5


/sig windows

**Comment by jsturtevant** [comment]
possibly related to https://github.com/containerd/ttrpc/issues/72 and https://github.com/containerd/ttrpc/pull/73 thought stack traces don't match directly

**Comment by jsturtevant** [comment]
Tested the above with the changes in https://github.com/containerd/ttrpc/pull/73 and there is no deadlock

**Comment by jsturtevant** [comment]
Fixed in https://github.com/containerd/ttrpc/pull/94

Waiting for fix to get into containerd

**Comment by jsturtevant** [comment]
fix was released in containerd 1.5.8 and 1.6.beta3
/close
